File: src/client/vad.js
/**
 * Client-Side Voice Activity Detection (VAD)
 * 
 * Detects when user starts and stops speaking using simple energy-based detection.
 * For production, consider using @ricky0123/vad-web or similar library.
 */

class ClientVAD {
  constructor(options = {}) {
    this.aggressiveness = options.aggressiveness || 3;
    this.energyThreshold = options.energyThreshold || 0.01; // RMS threshold
    this.silenceThreshold = options.silenceThreshold || 10; // frames
    this.speechThreshold = options.speechThreshold || 3; // frames
    
    // State
    this.isSpeaking = false;
    this.silenceFrames = 0;
    this.speechFrames = 0;
    
    console.log('[VAD] Initialized with aggressiveness:', this.aggressiveness);
  }
  
  /**
   * Process an audio chunk and detect speech events
   * 
   * @param {Float32Array} audioData - Audio samples normalized between -1 and 1
   * @returns {string} Event: 'START', 'SPEECH', 'END', or 'SILENCE'
   */
  processChunk(audioData) {
    const hasSpeech = this.detectSpeech(audioData);
    
    if (hasSpeech) {
      this.speechFrames++;
      this.silenceFrames = 0;
      
      if (!this.isSpeaking && this.speechFrames >= this.speechThreshold) {
        // Transition: SILENCE -> SPEECH
        this.isSpeaking = true;
        console.log('[VAD] Speech START detected');
        return 'START';
      } else if (this.isSpeaking) {
        return 'SPEECH';
      }
    } else {
      this.silenceFrames++;
      this.speechFrames = 0;
      
      if (this.isSpeaking && this.silenceFrames >= this.silenceThreshold) {
        // Transition: SPEECH -> SILENCE
        this.isSpeaking = false;
        console.log('[VAD] Speech END detected');
        return 'END';
      } else if (this.isSpeaking) {
        // Still in speech, but this chunk is silent
        return 'SPEECH';
      }
    }
    
    return 'SILENCE';
  }
  
  /**
   * Detect speech in audio chunk using RMS energy
   * 
   * @param {Float32Array} audioData - Audio samples
   * @returns {boolean} True if speech detected
   */
  detectSpeech(audioData) {
    if (!audioData || audioData.length === 0) {
      return false;
    }
    
    // Calculate RMS (Root Mean Square) energy
    const rms = this.calculateRMS(audioData);
    
    // Adjust threshold based on aggressiveness
    const threshold = this.energyThreshold * (4 - this.aggressiveness);
    
    return rms > threshold;
  }
  
  /**
   * Calculate RMS energy of audio samples
   * 
   * @param {Float32Array} audioData - Audio samples
   * @returns {number} RMS value
   */
  calculateRMS(audioData) {
    let sum = 0;
    for (let i = 0; i < audioData.length; i++) {
      sum += audioData[i] * audioData[i];
    }
    return Math.sqrt(sum / audioData.length);
  }
  
  /**
   * Reset VAD state
   */
  reset() {
    this.isSpeaking = false;
    this.silenceFrames = 0;
    this.speechFrames = 0;
    console.log('[VAD] State reset');
  }
  
  /**
   * Set energy threshold dynamically
   * 
   * @param {number} threshold - New threshold value
   */
  setThreshold(threshold) {
    this.energyThreshold = threshold;
    console.log('[VAD] Threshold updated to:', threshold);
  }
}

// For production, use a more sophisticated VAD library:
// 
// Option 1: @ricky0123/vad-web (Silero VAD in browser)
// import { MicVAD } from "@ricky0123/vad-web"
// const vad = await MicVAD.new({
//   onSpeechStart: () => console.log("Speech start"),
//   onSpeechEnd: (audio) => console.log("Speech end", audio),
// })
//
// Option 2: WebRTC VAD (if available via WASM)
// Option 3: ML-based VAD using TensorFlow.js

export default ClientVAD;

File: legacy/vad.py
"""
Voice Activity Detection (VAD) Module.

This module detects when a user starts and stops speaking.
"""

from enum import Enum


class VADEvent(Enum):
    """Events emitted by the VAD processor."""
    SILENCE = "SILENCE"
    START = "START"  # User starts speaking
    SPEECH = "SPEECH"  # User is speaking (continuation)
    END = "END"  # User stops speaking


class VADProcessor:
    """
    Voice Activity Detection processor.
    
    This is a stateful processor that analyzes incoming audio chunks
    and emits events when speech starts and ends.
    """
    
    def __init__(self, aggressiveness: int = 3):
        """
        Initialize the VAD processor.
        
        Args:
            aggressiveness: VAD sensitivity (0-3, where 3 is most aggressive)
        """
        self.aggressiveness = aggressiveness
        self.is_speaking = False
        self.silence_frames = 0
        self.speech_frames = 0
        
        # Thresholds for triggering events
        self.speech_threshold = 3  # Frames of speech to trigger START
        self.silence_threshold = 10  # Frames of silence to trigger END
        
        print(f"[VAD] Initialized with aggressiveness={aggressiveness}")
    
    def process_chunk(self, audio_chunk: bytes) -> VADEvent:
        """
        Process a single audio chunk and return the current VAD event.
        
        Args:
            audio_chunk: Raw audio bytes (PCM 16-bit, 16kHz recommended)
            
        Returns:
            VADEvent indicating the current speech state
        """
        # TODO: Replace this stub with actual VAD logic
        # Options:
        #   - webrtcvad library
        #   - silero-vad model
        #   - pyannote.audio
        #   - Custom energy-based detection
        
        # Stub implementation: detect speech based on audio energy
        has_speech = self._detect_speech_in_chunk(audio_chunk)
        
        if has_speech:
            self.speech_frames += 1
            self.silence_frames = 0
            
            if not self.is_speaking and self.speech_frames >= self.speech_threshold:
                # Transition: SILENCE -> SPEECH
                self.is_speaking = True
                print("[VAD] Speech START detected")
                return VADEvent.START
            elif self.is_speaking:
                return VADEvent.SPEECH
        else:
            self.silence_frames += 1
            self.speech_frames = 0
            
            if self.is_speaking and self.silence_frames >= self.silence_threshold:
                # Transition: SPEECH -> SILENCE
                self.is_speaking = False
                print("[VAD] Speech END detected")
                return VADEvent.END
            elif self.is_speaking:
                # Still in speech, but this chunk is silent
                return VADEvent.SPEECH
        
        return VADEvent.SILENCE
    
    def _detect_speech_in_chunk(self, audio_chunk: bytes) -> bool:
        """
        Simple energy-based speech detection (stub).
        
        Args:
            audio_chunk: Raw audio bytes
            
        Returns:
            True if speech is detected, False otherwise
        """
        # TODO: Implement proper VAD algorithm
        # For now, use simple energy threshold
        if len(audio_chunk) == 0:
            return False
        
        # Calculate RMS energy
        import struct
        samples = struct.unpack(f"{len(audio_chunk)//2}h", audio_chunk)
        rms = (sum(s*s for s in samples) / len(samples)) ** 0.5
        
        # Threshold (adjust based on your audio characteristics)
        threshold = 500
        return rms > threshold
    
    def reset(self):
        """Reset the VAD state."""
        self.is_speaking = False
        self.silence_frames = 0
        self.speech_frames = 0
        print("[VAD] State reset")

